osbs/utils.py

```diff
@@ -188,29 +191,36 @@ def buildconfig_update(orig, new, remove_nonexistent_keys=False):
 
 
 @contextlib.contextmanager
-def checkout_git_repo(git_uri, git_ref, git_branch=None):
+def checkout_git_repo(git_uri, git_ref, git_branch=None, max_retries=GIT_MAX_RETRIES,
+                      cooldown=GIT_BACKOFF_FACTOR):
     tmpdir = tempfile.mkdtemp()
     repo_path = os.path.join(tmpdir, "repo")
+    # when you clone into an empty directory and cloning process fails
+    # git will remove the empty directory (why?!)
     try:
-        try:
-            # when you clone into an empty directory and cloning process fails
-            # git will remove the empty directory (why?!)
-            args = ['git', 'clone', git_uri]
-            if git_branch:
-                args += ['-b', git_branch]
-
-            args.append(repo_path)
-            run_command(args)
-        except OsbsException as ex:
-            raise OsbsException("Unable to clone git repo '%s' "
-                                "branch '%s'" % (git_uri, git_branch),
-                                cause=ex, traceback=sys.exc_info()[2])
+        for counter in range(max_retries + 1):
```

Comment:
Could we separate retry logic from git clone logic here?

* construction of git checkout command is immutable and can be defined out of cycle

```
cmd = construct_git_checkout(...)
try:
    subprocess_call_with_retry(cmd, ....)  # emits subprocess.CalledProcessError when all tries failed
except subprocess.CalledProcessError as ex:
    raise OsbsException(...)
```